### PR TITLE
libgcc: Init at 7.3.0

### PR DESCRIPTION
**Pull request by Ericson2314**
###### Motivation for this change

I want to start from this, and work towards packaging all the libraries this way. Then we can:

1. Go from 2 cross compilers ("stage static" "stage final") to just 1
7. Easier LLVM cross compilation without needing to work around the 2 GCC dance.
2. Radically simplify CC wrapper (compiler-rt is already built separately)
3. Simplify native and cross boostrapping (as libraries and tools can always be built separately during bootstrapping)
4. Help with @angerman's windows improvements (where custom bootstrapping is needed as some more GCC libraries require `libc`).
5. Radically simplify GCC derivation. There might be net more code, as at least `libgcc` is hard to build on its own, but the each derivation will be much simpler
6. Build `libstdc++` separately for `libstdcxxClang`

This is analogous to @kirelagin and @matthewbauer's work rebuilding GHC wired-in libraries.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [x] cross ->linux
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/development/libraries/gcc/libgcc/default.nix [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+    export AS=${buildPackages.stdenv.cc}/bin/$AS
+    export CC=${buildPackages.stdenv.cc}/bin/$CC
+    export CPP=${buildPackages.stdenv.cc}/bin/$CPP
+    export CXX=${buildPackages.stdenv.cc}/bin/$CXX
+    export LD=${buildPackages.stdenv.cc.bintools}/bin/$LD
+
+    export AS_FOR_TARGET=${stdenvNoLibs.cc}/bin/$AS_FOR_TARGET
+    export CC_FOR_TARGET=${stdenvNoLibs.cc}/bin/$CC_FOR_TARGET
+    export CPP_FOR_TARGET=${stdenvNoLibs.cc}/bin/$CPP_FOR_TARGET
+    export LD_FOR_TARGET=${stdenvNoLibs.cc.bintools}/bin/$LD_FOR_TARGET
```
Comment (Ericson2314) [inline pkgs/development/libraries/gcc/libgcc/default.nix]: This and GHC show how we really need to change binutils-wrapper to make these full paths across the board, not just names on the PATH.

**Comment by angerman** [comment]
I did something similar in  #40637, the primary motivation was to build gcc+winpthreads to support c++11 threading features.  That however turned out to produce libraries that behaved subtly different from those built with the msvc. As such I did not persue that further. 

Building with libwpthreads requries building gcc (no libs), then winpthreads with that gcc, and finally the rest of the libs to package it all together. So this migth just work out right with this appraoch.

**Comment by Ericson2314** [comment]
Oh! I somehow missed that and assume you never upstreamed it. My bad!

**Comment by Ericson2314** [comment]
Are you saying that your old PR didn't work, but this might?

**Comment by angerman** [comment]
It's not upstreamed and it's not perfectly working. Yes. I think this might be a clearer approach; you've got way more experience with the low level details (doh!).

But maybe there's a nugget or two you can use :-)

**Comment by Ericson2314** [comment]
Well this wouldn't fix any C++ ABI issues with GCC, but it might make cross with LLVM easier, and LLVM on/for Windows has recently seen a lot of work.

**Pull request merged**